WinRT.Runtime/ObjectReference.cs

```diff
@@ -197,25 +184,34 @@ public static ObjectReference<T> FromAbi(IntPtr thisPtr)
             {
                 return null;
             }
-            var (vftblIUnknown, vftblT) = GetVtables(thisPtr);
-            return FromAbi(thisPtr, vftblIUnknown, vftblT);
+            var vftblT = GetVtable(thisPtr);
+            return FromAbi(thisPtr, vftblT);
         }
 
-        // C# doesn't allow us to express that T contains IUnknownVftbl, so we'll use a tuple
-        private static unsafe (IUnknownVftbl vftblIUnknown, T vftblT) GetVtables(IntPtr thisPtr)
+        private static unsafe T GetVtable(IntPtr thisPtr)
         {
             var vftblPtr = Unsafe.AsRef<VftblPtr>(thisPtr.ToPointer());
-            var vftblIUnknown = Marshal.PtrToStructure<IUnknownVftbl>(vftblPtr.Vftbl);
             T vftblT;
+            // With our vtable types, the generic vtables will have System.Delegate fields
+            // and the non-generic types will have only void* fields.
+            // On .NET 5, we can use RuntimeHelpers.IsReferenceorContainsReferences
+            // to disambiguate between generic and non-generic vtables since it's a JIT-time constant.
+            // Since it is a JIT time constant, this function will be branchless on .NET 5.
+            // On .NET Standard 2.0, the IsReferenceOrContainsReferences method does not exist,
+            // so we instead fall back to typeof(T).IsGenericType, which sadly is not a JIT-time constant.
+#if NETSTANDARD2_0
             if (typeof(T).IsGenericType)
+#else
+            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+#endif
             {
                 vftblT = (T)typeof(T).GetConstructor(BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.CreateInstance, null, new[] { typeof(IntPtr) }, null).Invoke(new object[] { thisPtr });
             }
             else
             {
-                vftblT = Marshal.PtrToStructure<T>(vftblPtr.Vftbl);
+                vftblT = Unsafe.AsRef<T>(vftblPtr.Vftbl.ToPointer());
```

Comment:
is AsRef more performant that PtrToStructure?

Reply:
Significantly. The JIT can see that `Unsafe.AsRef` is a pointer dereference instead of having to get the type and calling into the runtime and doing a copy.